mesonpy/__init__.py

```diff
@@ -685,6 +686,44 @@ def _strings(value: Any, name: str) -> List[str]:
     return scheme(table, 'tool.meson-python')
 
 
+def _validate_config_settings(config_settings: Dict[str, Any]) -> Dict[str, Any]:
+    """Validate options received from build frontend."""
+
+    def _string(value: Any, name: str) -> str:
+        if not isinstance(value, str):
+            raise ConfigError(f'only one value for "{name}" can be specified')
```

Comment:
This error messages assumes it can only be a container, while in fact it can be anything. For most PEP 517 frontends, it will be a container of some sort, but that is not guaranteed.

Reply:
No, it can be only a string or a list of strings. It is written in the PEP

Reply:
The PEP only makes a *recommendation* when frontends are building the config settings from a user provided string mapping :sweat_smile:

In the PEP, you can see 3 words capitalized, "MUST", "SHOULD", and "MAY", they indicate obligation, recommendation, and optionality, respectively. So while the PEP does recommend for that to be the case, it does not mandate it. It also explicitly mentions that fronts can have an alternative mechanism to build the dictionary.

> Build frontends MAY also provide arbitrary other mechanisms for users to place entries in this dictionary.

It's a simple enough check anyway.

Reply:
Please, if you need to quote from the PEP, do no quote only the part that makes it look like it supports your claims:

> This argument, which is passed to all hooks, is an arbitrary dictionary provided as an “escape hatch” for users to pass ad-hoc configuration into individual package builds. Build backends MAY assign any semantics they like to this dictionary. Build frontends SHOULD provide some mechanism for users to specify arbitrary string-key/string-value pairs to be placed in this dictionary. For example, they might support some syntax like --package-config CC=gcc. In case a user provides duplicate string-keys, build frontends SHOULD combine the corresponding string-values into a list of strings. Build frontends MAY also provide arbitrary other mechanisms for users to place entries in this dictionary.

The dictionary is string-key, string-value. You can use arbitrary mechanism to add entries to the dictionary, but the dictionary has string keys and string values.

Reply:
Can you please take your time and read all the text?

> Build frontends SHOULD provide some mechanism for users to specify arbitrary string-key/string-value pairs to be placed in this dictionary.

Reply:
https://github.com/pypa/setuptools/blob/main/setuptools/build_meta.py#L142